### ISSUE TITLE
Issues with abstract-method with methods from libraries

### ISSUE THREAD

**Issue by kasium** [PyCQA/pylint#5031]
### Bug description

```python
# Original this issue comes from sqlalchemy (1.4.22)

from sqlalchemy.ext.compiler import compiles
from sqlalchemy.sql.functions import ReturnTypeFromArgs

class greatest(ReturnTypeFromArgs):
    name = "greatest"


@compiles(greatest, "foo")
def hana_greatest(element, compiler, **_):
    return compiler.visit_function(element)


# More simple

class TheInternalClass:

    def do(self):
        return 42

    def _some_internal(self):
        raise NotImplementedError

class TheImplementedClass(TheInternalClass):

    def do_more(self):
        return self.do() * 2
```

### Command used

```shell
pylint dummy.py --disable all --enable abstract-method,useless-suppression
```

### Pylint output

```shell
dummy.py:9:0: W0223: Method '_some_internal' is abstract in class 'TheInternalClass' but is not overridden (abstract-method)
```

### Expected behavior

So, pylint can obviously not figure out, that _some_internal is by purpose not implemented and should **not** be implemented by an user (assuming that `TheInternalClass` is part of a library).

So, the abstract find logic is implemented in astroid: https://github.com/PyCQA/astroid/blob/a59b1cf97a5da7288d8e71791d6b2723df84cc11/astroid/nodes/scoped_nodes.py#L1706

It might be an idea to add a configuration which tells pylint which methods are abstract by purpose

```ini
abstract-methods =
    TheInternalClass._some_internal
```

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.0
Python 3.7.1 (default, Oct 22 2018, 13:16:18) [GCC]
```

### OS / Environment

Linux, SUSE

**Comment by Pierre-Sassoulas** [comment]
Hello, I'm not certain of it but I'm think that if the internal class is declared as abstract (inherit from ABC) the warning should disapear.

**Comment by Pierre-Sassoulas** [comment]
Just checked that this is working as intended when you use ``abc.ABC`` to indicate that the class is abstract.

**Comment by kasium** [comment]
Thanks a lot

**Comment by zzzeek** [comment]
hey there -

whats the style guide which states that classes that dont use Python abc are not allowed to use NotImplementedError?  How can pylint claim a class is "abstract" when it is not using abc at all?  seems pretty wrong to me so far, can i be enlightened?

**Comment by Pierre-Sassoulas** [comment]
Hey, the ``abc`` module is the official way to declare a class as abstract, since its introduction. Before that there were ``raise NotImplementedError`` to denote abstractness. That's why pylint detect such a class as abstract. The rational for using abc is in [PEP3119](https://www.python.org/dev/peps/pep-3119/),  you can also check how it was used for numbers in [PEP3141](https://www.python.org/dev/peps/pep-3141/). But the main advantages for me is explicitness and enforced abstractness during initialization. There will be a an explicit fail when instancing an abstract class if a function is not implemented instead of failing only when accessing the unimplemented function, so it enforces a better, clearer design directly handled by the interpreter without relying on triggering the error by running the code.

**Comment by zzzeek** [comment]
yes I know all about abc, thanks!  that's not the issue here.  the issue is one of my users is trying to run pylint and for some reason it's digging into sqlalchemy and somehow complaining about this.    if pyllint cannot correctly detect when an abstract method is correctly implemented, for some value of "abstract" (in this case, "saw notimplemented, therefore this class is abstract"), that's still a bug in pylint.    neither of the two PEPs you link state anything about "it is now incorrect to use NotImplementedError for a class that is not making use of abc".   SQLAlchemy 1.4 has a variety of reasons it cant use "abc" for all classes everywhere.    user here should not be presented with pylint complaints about an upstream library in any case.

**Comment by Pierre-Sassoulas** [comment]
>  if pyllint cannot correctly detect when an abstract method is correctly implemented, for some value of "abstract" (in this case, "saw notimplemented, therefore this class is abstract"), that's still a bug in pylint. 

Well yes, if the function is correctly implemented it's a bug in pylint (well astroid) not detecting the concrete implementation. I somehow missed that the ``# more simple`` example was not the whole picture of this issue.

**Comment by zzzeek** [comment]
it detects it here so i will resolve in that manner, thanks!
